File: kotlin/advent/src/main/kotlin/org/gristle/pdxpuzzles/advent/y2024/Y24D13.kt
package org.gristle.pdxpuzzles.advent.y2024

import org.gristle.pdxpuzzles.advent.utilities.Day
import org.gristle.pdxpuzzles.utilities.parsing.getLongs

class Y24D13(input: String) : Day {
    data class ClawMachine(
        val a1: Long,
        val a2: Long,
        val b1: Long,
        val b2: Long,
        val c1: Long,
        val c2: Long,
    ) {
        fun minimumTokens(): Pair<Long, Long>? {
            // use elimination method of solving system of equations. Using these equations:
            //   1) a1x + b1y = c1
            //   2) a2x + b2y = c2
            // we'll find the lcm of a1 and a2. then we'll multiply equation 1 so that a1 is the lcm,
            // and we'll multiply equation 2 so that a2 is -lcm. Then add the two equations together,
            // which gives all we need to solve for y
            val commonMultiple = a1 * a2
            val f1 = commonMultiple / a1
            val f2 = -(commonMultiple / a2)
            val yNumerator = (c1 * f1 + c2 * f2)
            val yDenominator = (b1 * f1 + b2 * f2)
            val y = yNumerator / yDenominator
            if (y * yDenominator != yNumerator) return null
            val xNumerator = (c1 - b1 * y)
            val xDenominator = a1
            val x = xNumerator / xDenominator
            if (x * xDenominator != xNumerator) return null
            return x to y
        }
    }

    private val machines = input
        .getLongs()
        .chunked(6)
        .map { ClawMachine(it[0], it[1], it[2], it[3], it[4], it[5]) }
        .toList()

    override fun part1() = machines
        .mapNotNull { it.minimumTokens() }
        .filter { (a, b) -> a <= 100 && b <= 100 }
        .sumOf { (a, b) -> a * 3 + b }
    override fun part2() = machines
        .map { it.copy(c1 = it.c1 + 10000000000000L, c2 = it.c2 + 10000000000000L) }
        .mapNotNull { it.minimumTokens() }
        .sumOf { (a, b) -> a * 3 + b }
}

fun main() = Day.runDay(Y24D13::class)

//    Class creation: 8ms
//    Part 1: 37128 (4ms)
//    Part 2: 74914228471331 (4ms)
//    Total time: 17ms

private val test = listOf("""Button A: X+94, Y+34
Button B: X+22, Y+67
Prize: X=8400, Y=5400

Button A: X+26, Y+66
Button B: X+67, Y+21
Prize: X=12748, Y=12176

Button A: X+17, Y+86
Button B: X+84, Y+37
Prize: X=7870, Y=6450

Button A: X+69, Y+23
Button B: X+27, Y+71
Prize: X=18641, Y=10279
""")

File: rust/advent/src/bin/2024/y24d13.rs
use advent::utilities::get_input::get_input;
use itertools::Itertools;
use utilities::{
    parsing::get_numbers::ContainsNumbers,
    structs::stopwatch::{ReportDuration, Stopwatch},
};

type Input<'a> = &'a str;
type Output = i64;

fn main() {
    let mut stopwatch = Stopwatch::new();
    stopwatch.start();
    let input = get_input(24, 13).unwrap();
    println!("Input parsed ({})", stopwatch.lap().report());
    println!("1. {} ({})", part1(&input), stopwatch.lap().report());
    println!("2. {} ({})", part2(&input), stopwatch.lap().report());
    println!("Total: {}", stopwatch.stop().report());
}

fn solve(input: Input, additional: i64) -> Output {
    input
        .get_numbers()
        .tuples()
        .filter_map(|(a1, a2, b1, b2, c1, c2)| {
            let c1 = c1 + additional;
            let c2 = c2 + additional;
            tokens(a1, a2, b1, b2, c1, c2)
        })
        .map(|(a, b)| a * 3 + b)
        .sum()
}

fn tokens(a1: i64, a2: i64, b1: i64, b2: i64, c1: i64, c2: i64) -> Option<(i64, i64)> {
    // use elimination method of solving system of equations. Using these equations:
    //   1) a1x + b1y = c1
    //   2) a2x + b2y = c2
    // we'll find the lcm of a1 and a2. then we'll multiply equation 1 so that a1 is the lcm,
    // and we'll multiply equation 2 so that a2 is -lcm. Then add the two equations together,
    // which gives all we need to solve for y
    let common_multiple = a1 * a2;
    let f1 = common_multiple / a1;
    let f2 = -common_multiple / a2;
    let y_numerator = c1 * f1 + c2 * f2;
    let y_denominator = b1 * f1 + b2 * f2;
    let y = y_numerator / y_denominator;
    if y * y_denominator != y_numerator {
        return None;
    }
    let x_numerator = c1 - b1 * y;
    let x_denominator = a1;
    let x = x_numerator / x_denominator;
    if x * x_denominator != x_numerator {
        return None;
    }
    Some((x, y))
}

fn part1(input: Input) -> Output {
    solve(input, 0)
}
fn part2(input: Input) -> Output {
    solve(input, 10_000_000_000_000)
}

#[test]
fn default() {
    let input = get_input(24, 13).unwrap();
    assert_eq!(37128, part1(&input));
    assert_eq!(74914228471331, part2(&input));
}

// Input parsed (27μs)
// 1. 37128 (69μs)
// 2. 74914228471331 (65μs)
// Total: 165μs
